docs/design/llvm_sparse_runtime.md

```diff
@@ -0,0 +1,458 @@
+                                                                 |               |            |
+                                                                 |               +------------+
+                                                                 |
+                                                                 +-------------> +------------+
+                                                                                 |            |
+                                                                                 |   cell-1   |
+                                                                                 |            |
+                                                                                 +------------+
+```
+
+We can follow [`Pointer_activate`](https://github.com/taichi-dev/taichi/blob/0f4fb9c662e6e3ffacc26e7373258d8d0414423b/taichi/runtime/llvm/node_pointer.h#L41-L65) to see how `pointer` SNode is implemented using the sparse runtime infrastructure.
+
+```cpp
+void Pointer_activate(Ptr meta_, Ptr node, int i) {
+  auto meta = (StructMeta *)meta_;
+  auto num_elements = Pointer_get_num_elements(meta_, node);
+  // 1
+  volatile Ptr lock = node + 8 * i;
+  volatile Ptr *data_ptr = (Ptr *)(node + 8 * (num_elements + i));
+  // 2
+  if (*data_ptr == nullptr) {
+    // 3
+    // The cuda_ calls will return 0 or do noop on CPUs
+    u32 mask = cuda_active_mask();
+    if (is_representative(mask, (u64)lock)) {
+      // 4
+      locked_task(
+          lock,
+          [&] {
+            // 5
+            auto rt = meta->context->runtime;
+            auto alloc = rt->node_allocators[meta->snode_id];
+            auto allocated = (u64)alloc->allocate();
+            atomic_exchange_u64((u64 *)data_ptr, allocated);
+          },
+          [&]() { return *data_ptr == nullptr; });
+    }
+    warp_barrier(mask);
+  }
+}
+```
+
+1. Retrieves both the lock and the pointer for the `i`-th cell. Note that the pointer width is assumed to be 8-byte wide. Locks are simple 64-bit integers. According to the layout, there are `max_num_elements` number of locks, followed by `max_num_elements` number of pointer to cells.
+2. Checks whether the content of `data_ptr` is `nullptr` without any lock. This is the classical [double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking) pattern.
+3. If 2 is true, pick one thread within a CUDA warp to acquire the lock. This is a small optimization to prevent lock contention.
+4. The winning thread tries to acquire the lock using [`locked_task`](https://github.com/taichi-dev/taichi/blob/master/taichi/runtime/llvm/locked_task.h).
+5. Retrieves the memory allocator (`node_allocators`) for this particular SNode, allocates a new cell, and stores the address of the allocated cell into `data_pointer`. Because the cell size of each SNode is different, the runtime has a dedicated allocator for each SNode, which knows how much space to allocate per cell.
```

Comment:
```suggestion
5. Retrieves the memory allocator (`node_allocators`) for this particular SNode, allocates a new cell, and stores the address of the allocated cell into `data_ptr`. Because the cell size of each SNode is different, the runtime has a dedicated allocator for each SNode, which knows how much space to allocate per cell.
```